File: udon_decompiler/analysis/transform/pass_base.py
from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, Iterable, List, Optional, Sequence, cast

from udon_decompiler.analysis.ir.control_flow_graph import ControlFlowGraph
from udon_decompiler.analysis.ir.control_flow_node import ControlFlowNode
from udon_decompiler.analysis.ir.nodes import (
    IRBlock,
    IRClass,
    IRFunction,
)
from udon_decompiler.analysis.transform.ir_utils import iter_block_containers
from udon_decompiler.models.program import UdonProgramData


class TransformStepper:
    """No-op stepper"""

    def start_group(self, _description: str) -> None:
        return

    def end_group(self, keep_if_empty: bool = False) -> None:
        _ = keep_if_empty
        return

    def step(self, _description: str) -> None:
        return


@dataclass
class ProgramTransformContext:
    """
    Program-level transform context.
    """

    program: UdonProgramData
    ir_class: Optional[IRClass] = None
    settings: Dict[str, Any] = field(default_factory=dict)
    metadata: Dict[str, Any] = field(default_factory=dict)
    stepper: TransformStepper = field(default_factory=TransformStepper)
    cancellation_check: Optional[Callable[[], None]] = None

    def throw_if_cancellation_requested(self) -> None:
        if self.cancellation_check is not None:
            self.cancellation_check()

    def create_il_context(self, function: IRFunction) -> "TransformContext":
        return TransformContext(function=function, program_context=self)


class TransformContext:
    """Per-function context for ITransform"""

    def __init__(
        self,
        function: IRFunction,
        program_context: ProgramTransformContext,
    ):
        self.function = function
        self.program_context = program_context

    @property
    def program(self) -> UdonProgramData:
        return self.program_context.program

    @property
    def ir_class(self) -> Optional[IRClass]:
        return self.program_context.ir_class

    @ir_class.setter
    def ir_class(self, value: Optional[IRClass]) -> None:
        self.program_context.ir_class = value

    @property
    def settings(self) -> Dict[str, Any]:
        return self.program_context.settings

    @property
    def metadata(self) -> Dict[str, Any]:
        return self.program_context.metadata

    @property
    def stepper(self) -> TransformStepper:
        return self.program_context.stepper

    def throw_if_cancellation_requested(self) -> None:
        self.program_context.throw_if_cancellation_requested()

    def step_start_group(self, description: str) -> None:
        self.stepper.start_group(description)

    def step_end_group(self, keep_if_empty: bool = False) -> None:
        self.stepper.end_group(keep_if_empty=keep_if_empty)

    def step(self, description: str) -> None:
        self.stepper.step(description)


class IProgramTransform(ABC):
    """Program-level transform that sees all functions."""

    name: Optional[str] = None

    @property
    def display_name(self) -> str:
        return self.name or self.__class__.__name__

    @abstractmethod
    def run(
        self,
        functions: List[IRFunction],
        context: ProgramTransformContext,
    ) -> None:
        raise NotImplementedError


class ITransform(ABC):
    """Per-function transform."""

    name: Optional[str] = None

    @property
    def display_name(self) -> str:
        return self.name or self.__class__.__name__

    @abstractmethod
    def run(self, function: IRFunction, context: TransformContext) -> None:
        raise NotImplementedError


class IBlockTransform(ABC):
    """Per-block transform, usually orchestrated by BlockTransform."""

    @abstractmethod
    def run(self, block: IRBlock, context: "BlockTransformContext") -> None:
        raise NotImplementedError


class IStatementTransform(ABC):
    """Per-statement transform, executed right-to-left inside one block."""

    @abstractmethod
    def run(
        self,
        block: IRBlock,
        pos: int,
        context: "StatementTransformContext",
    ) -> None:
        raise NotImplementedError


class BlockTransformContext(TransformContext):
    def __init__(
        self,
        context: TransformContext,
        control_flow_graph: ControlFlowGraph,
    ):
        super().__init__(
            function=context.function,
            program_context=context.program_context,
        )
        self.container = control_flow_graph.container
        self.control_flow_graph = control_flow_graph

        self.block: Optional[IRBlock] = None
        self.control_flow_node: Optional[ControlFlowNode] = None
        self._dirty = False

    @property
    def is_dirty(self) -> bool:
        return self._dirty

    def mark_dirty(self) -> None:
        self._dirty = True

    def reset_dirty(self) -> None:
        self._dirty = False


class StatementTransformContext(TransformContext):
    def __init__(self, block_context: BlockTransformContext):
        super().__init__(
            function=block_context.function,
            program_context=block_context.program_context,
        )
        self.block_context = block_context

        self.rerun_current_position = False
        self.rerun_position: Optional[int] = None

    @property
    def block(self) -> Optional[IRBlock]:
        return self.block_context.block

    def request_rerun(self, pos: Optional[int] = None) -> None:
        if pos is None:
            self.rerun_current_position = True
            return
        if self.rerun_position is None or pos > self.rerun_position:
            self.rerun_position = pos


def run_block_transforms(
    block: IRBlock,
    transforms: Iterable[IBlockTransform],
    context: BlockTransformContext,
) -> None:
    for transform in transforms:
        context.throw_if_cancellation_requested()
        context.step_start_group(transform.__class__.__name__)
        transform.run(block, context)
        context.step_end_group()


def run_il_transforms(
    function: IRFunction,
    transforms: Iterable[ITransform],
    context: TransformContext,
) -> None:
    for transform in transforms:
        context.throw_if_cancellation_requested()
        context.step_start_group(transform.display_name)
        transform.run(function, context)
        context.step_end_group(keep_if_empty=True)


class LoopingBlockTransform(IBlockTransform):
    """
    Repeats child block transforms until the block no longer changes.
    """

    def __init__(self, *transforms: IBlockTransform):
        self.transforms: Sequence[IBlockTransform] = transforms
        self._running = False

    def run(self, block: IRBlock, context: BlockTransformContext) -> None:
        if self._running:
            raise RuntimeError(
                "LoopingBlockTransform already running. Transforms are not re-entrant."
            )

        self._running = True
        try:
            count = 1
            while True:
                context.reset_dirty()
                run_block_transforms(block, self.transforms, context)
                if not context.is_dirty:
                    break
                count += 1
                context.step(f"Block is dirty; running loop iteration #{count}.")
        finally:
            self._running = False


class StatementTransform(IBlockTransform):
    """
    Runs statement transforms with right-to-left + rerun semantics.
    """

    def __init__(self, *children: IStatementTransform):
        self.children: Sequence[IStatementTransform] = children

    def run(self, block: IRBlock, context: BlockTransformContext) -> None:
        if not block.statements:
            return

        stmt_ctx = StatementTransformContext(context)

        pos = 0
        stmt_ctx.rerun_position = len(block.statements) - 1
        while pos >= 0:
            context.throw_if_cancellation_requested()
            if not block.statements:
                break

            if stmt_ctx.rerun_position is not None:
                pos = min(stmt_ctx.rerun_position, len(block.statements) - 1)
                stmt_ctx.rerun_position = None
            elif pos >= len(block.statements):
                pos = len(block.statements) - 1

            for transform in self.children:
                transform.run(block, pos, stmt_ctx)

                if stmt_ctx.rerun_current_position:
                    stmt_ctx.rerun_current_position = False
                    stmt_ctx.request_rerun(pos)
                if stmt_ctx.rerun_position is not None:
                    break

            if stmt_ctx.rerun_position is None:
                pos -= 1


class BlockTransform(ITransform):
    """
    Transform that runs block transforms in dominator-tree order.

    Pre-order transforms run before dominated children;
    post-order transforms run after children.
    """

    def __init__(self):
        self.pre_order_transforms: List[IBlockTransform] = []
        self.post_order_transforms: List[IBlockTransform] = []
        self._running = False

    def run(self, function: IRFunction, context: TransformContext) -> None:
        if self._running:
            raise RuntimeError(
                "Reentrancy detected. Transforms are not thread-safe/re-entrant."
            )

        self._running = True
        try:
            # Snapshot containers before transforms
            # This avoids re-processing newly created containers
            # in the same BlockTransform run (e.g. during loop/switch detection).
            containers = list(iter_block_containers(function))
            for container in containers:
                context.throw_if_cancellation_requested()
                if container.entry_block is None:
                    continue

                cfg = ControlFlowGraph(
                    container=container,
                    function_body=function.body,
                )
                entry_node = cfg.get_node(container.entry_block)

                block_ctx = BlockTransformContext(
                    context=context,
                    control_flow_graph=cfg,
                )

                self._visit_block(entry_node, block_ctx)
        finally:
            self._running = False

    def _visit_block(
        self,
        cfg_node: ControlFlowNode,
        context: BlockTransformContext,
    ) -> None:
        block = cfg_node.block
        if block is None:
            raise RuntimeError("ControlFlowNode.block is None for non-sentinel node")
        context.block = block
        context.control_flow_node = cfg_node
        context.step_start_group(block.label)

        run_block_transforms(block, self.pre_order_transforms, context)

        for child in cfg_node.dominator_tree_children or []:
            self._visit_block(child, context)

        context.block = block
        context.control_flow_node = cfg_node
        run_block_transforms(block, self.post_order_transforms, context)

        context.step_end_group()

File: src/decompiler/transform/pass_base.rs
use std::collections::HashMap;

use crate::decompiler::context::DecompileContext;
use crate::decompiler::ir::{ControlFlowGraph, IrBlock, IrClass, IrFunction};
use crate::decompiler::{DecompileError, Result};

use super::ir_utils::{find_container_mut, iter_block_containers};

#[derive(Debug, Clone)]
pub struct ProgramTransformContext<'ctx> {
    pub class_name: String,
    pub ir_class: Option<IrClass>,
    pub decompile_context: &'ctx DecompileContext,
    pub settings: HashMap<String, String>,
    pub metadata: HashMap<String, i64>,
}

impl<'ctx> ProgramTransformContext<'ctx> {
    pub fn new(class_name: String, decompile_context: &'ctx DecompileContext) -> Self {
        Self {
            class_name,
            ir_class: None,
            decompile_context,
            settings: HashMap::new(),
            metadata: HashMap::new(),
        }
    }

    pub fn throw_if_cancellation_requested(&self) {}

    pub fn create_il_context(&mut self, function_name: String) -> TransformContext<'_, 'ctx> {
        TransformContext {
            function_name,
            program_context: self,
        }
    }
}

pub struct TransformContext<'a, 'ctx> {
    pub function_name: String,
    pub program_context: &'a mut ProgramTransformContext<'ctx>,
}

impl<'a, 'ctx> TransformContext<'a, 'ctx> {
    pub fn throw_if_cancellation_requested(&self) {
        self.program_context.throw_if_cancellation_requested();
    }
}

pub trait IProgramTransform {
    fn run(
        &self,
        functions: &mut [IrFunction],
        context: &mut ProgramTransformContext<'_>,
    ) -> Result<()>;
}

pub trait ITransform {
    fn run(&self, function: &mut IrFunction, context: &mut TransformContext<'_, '_>) -> Result<()>;
}

pub trait IBlockTransform {
    fn run(
        &self,
        block: &mut IrBlock,
        context: &mut BlockTransformContext<'_, '_, '_>,
    ) -> Result<()>;
}

pub trait IStatementTransform {
    fn run(
        &self,
        block: &mut IrBlock,
        pos: usize,
        context: &mut StatementTransformContext<'_, '_, '_, '_>,
    ) -> Result<()>;
}

pub struct BlockTransformContext<'a, 'b, 'ctx> {
    pub transform_context: &'a mut TransformContext<'b, 'ctx>,
    pub container_id: u32,
    pub control_flow_graph: ControlFlowGraph,
    pub block_index: Option<usize>,
    pub control_flow_node_index: Option<usize>,
    dirty: bool,
}

impl<'a, 'b, 'ctx> BlockTransformContext<'a, 'b, 'ctx> {
    fn new(
        transform_context: &'a mut TransformContext<'b, 'ctx>,
        container_id: u32,
        control_flow_graph: ControlFlowGraph,
    ) -> Self {
        Self {
            transform_context,
            container_id,
            control_flow_graph,
            block_index: None,
            control_flow_node_index: None,
            dirty: false,
        }
    }

    pub fn is_dirty(&self) -> bool {
        self.dirty
    }

    pub fn mark_dirty(&mut self) {
        self.dirty = true;
    }

    pub fn reset_dirty(&mut self) {
        self.dirty = false;
    }
}

pub struct StatementTransformContext<'a, 'b, 'c, 'ctx> {
    pub block_context: &'a mut BlockTransformContext<'b, 'c, 'ctx>,
    pub rerun_current_position: bool,
    pub rerun_position: Option<usize>,
}

impl<'a, 'b, 'c, 'ctx> StatementTransformContext<'a, 'b, 'c, 'ctx> {
    fn new(block_context: &'a mut BlockTransformContext<'b, 'c, 'ctx>) -> Self {
        Self {
            block_context,
            rerun_current_position: false,
            rerun_position: None,
        }
    }

    pub fn request_rerun(&mut self, pos: Option<usize>) {
        let Some(pos) = pos else {
            self.rerun_current_position = true;
            return;
        };
        if self.rerun_position.is_none_or(|x| pos > x) {
            self.rerun_position = Some(pos);
        }
    }
}

pub fn run_il_transforms(
    function: &mut IrFunction,
    transforms: &[Box<dyn ITransform>],
    program_context: &mut ProgramTransformContext<'_>,
) -> Result<()> {
    for transform in transforms {
        program_context.throw_if_cancellation_requested();
        let function_name = function.function_name.clone();
        let mut il_context = program_context.create_il_context(function_name);
        transform.run(function, &mut il_context)?;
    }
    Ok(())
}

pub fn run_block_transforms(
    block: &mut IrBlock,
    transforms: &[Box<dyn IBlockTransform>],
    context: &mut BlockTransformContext<'_, '_, '_>,
) -> Result<()> {
    for transform in transforms {
        context
            .transform_context
            .program_context
            .throw_if_cancellation_requested();
        transform.run(block, context)?;
    }
    Ok(())
}

pub struct LoopingBlockTransform {
    transforms: Vec<Box<dyn IBlockTransform>>,
}

impl LoopingBlockTransform {
    pub fn new(transforms: Vec<Box<dyn IBlockTransform>>) -> Self {
        Self { transforms }
    }
}

impl IBlockTransform for LoopingBlockTransform {
    fn run(
        &self,
        block: &mut IrBlock,
        context: &mut BlockTransformContext<'_, '_, '_>,
    ) -> Result<()> {
        loop {
            context.reset_dirty();
            run_block_transforms(block, &self.transforms, context)?;
            if !context.is_dirty() {
                break;
            }
        }
        Ok(())
    }
}

pub struct StatementTransform {
    children: Vec<Box<dyn IStatementTransform>>,
}

impl StatementTransform {
    pub fn new(children: Vec<Box<dyn IStatementTransform>>) -> Self {
        Self { children }
    }
}

impl IBlockTransform for StatementTransform {
    fn run(
        &self,
        block: &mut IrBlock,
        context: &mut BlockTransformContext<'_, '_, '_>,
    ) -> Result<()> {
        if block.statements.is_empty() {
            return Ok(());
        }

        let mut statement_context = StatementTransformContext::new(context);
        let mut pos = block.statements.len() - 1;
        statement_context.rerun_position = Some(pos);

        loop {
            if block.statements.is_empty() {
                break;
            }
            if let Some(rerun_position) = statement_context.rerun_position.take() {
                pos = rerun_position.min(block.statements.len() - 1);
            } else if pos >= block.statements.len() {
                pos = block.statements.len() - 1;
            }

            for transform in &self.children {
                transform.run(block, pos, &mut statement_context)?;
                if statement_context.rerun_current_position {
                    statement_context.rerun_current_position = false;
                    statement_context.request_rerun(Some(pos));
                }
                if statement_context.rerun_position.is_some() {
                    break;
                }
            }

            if statement_context.rerun_position.is_none() {
                if pos == 0 {
                    break;
                }
                pos -= 1;
            }
        }

        Ok(())
    }
}

pub struct BlockTransform {
    pub pre_order_transforms: Vec<Box<dyn IBlockTransform>>,
    pub post_order_transforms: Vec<Box<dyn IBlockTransform>>,
}

impl BlockTransform {
    pub fn new() -> Self {
        Self {
            pre_order_transforms: Vec::new(),
            post_order_transforms: Vec::new(),
        }
    }
}

impl Default for BlockTransform {
    fn default() -> Self {
        Self::new()
    }
}

impl ITransform for BlockTransform {

    fn run(&self, function: &mut IrFunction, context: &mut TransformContext<'_, '_>) -> Result<()> {
        let container_ids = iter_block_containers(function);

        for container_id in container_ids {
            let function_body_container_id = function.body.id;
            let Some(container) = find_container_mut(&mut function.body, container_id) else {
                continue;
            };

            let cfg = ControlFlowGraph::new(container, function_body_container_id);
            if cfg.nodes.is_empty() {
                continue;
            }

            let mut block_context = BlockTransformContext::new(context, container_id, cfg);
            run_dominator_tree_node(
                0,
                container,
                &self.pre_order_transforms,
                &self.post_order_transforms,
                &mut block_context,
            )?;
        }

        Ok(())
    }
}

fn run_dominator_tree_node(
    node_index: usize,
    container: &mut crate::decompiler::ir::IrBlockContainer,
    pre_transforms: &[Box<dyn IBlockTransform>],
    post_transforms: &[Box<dyn IBlockTransform>],
    context: &mut BlockTransformContext<'_, '_, '_>,
) -> Result<()> {
    if node_index >= container.blocks.len() {
        return Err(DecompileError::new(format!(
            "Dominator node index {} out of bounds for container {}",
            node_index, container.id
        )));
    }

    context.block_index = Some(node_index);
    context.control_flow_node_index = Some(node_index);

    {
        let block = container
            .blocks
            .get_mut(node_index)
            .ok_or_else(|| DecompileError::new("missing block for dominator node"))?;
        run_block_transforms(block, pre_transforms, context)?;
    }

    let children = context.control_flow_graph.nodes[node_index]
        .dominator_tree_children
        .clone()
        .unwrap_or_default();
    for child in children {
        run_dominator_tree_node(child, container, pre_transforms, post_transforms, context)?;
    }

    {
        let block = container
            .blocks
            .get_mut(node_index)
            .ok_or_else(|| DecompileError::new("missing block for dominator node"))?;
        run_block_transforms(block, post_transforms, context)?;
    }

    Ok(())
}
